native/android/SwrveSDKUnityBridge/src/main/java/com/swrve/sdk/SwrveUnityCommon.java

```diff
@@ -0,0 +1,392 @@
+{
+    public final static String SWRVE_TEMPORARY_PATH_KEY = "swrveTemporaryPath";
+    public final static String SDK_VERSION_KEY = "sdkVersion";
+    public final static String PREFAB_NAME_KEY = "prefabName";
+    public final static String DEVICE_INFO_KEY = "deviceInfo";
+
+    private final static String LOG_TAG = "UnitySwrveCommon";
+
+    private Map<String, Object> currentDetails;
+    protected WeakReference<Context> context;
+
+    private String sessionKey;
+
+    private File cacheDir;
+
+    private static SwrveUnityCommon instance;
+
+    public static boolean isInitialised() {
+        return null != instance;
+    }
+
+    public SwrveUnityCommon() { }
+
+    public void init(String jsonString) {
+        init(
+            UnityPlayer.currentActivity.getApplicationContext(),
+            jsonString
+        );
+    }
+
+    /***
+     * This is the most proper Constructor to call from the Unity layer,
+     * via a native plugin, with a jsonString of settings, which will be
+     * then cached locally.
+     *
+     * @param context ApplicationContext
+     * @param jsonString JSON String of configuration
+     */
+    public void init(Context context, String jsonString) {
+        SwrveLogger.d(
+            "UnitySwrveCommon constructor called with jsonString" +
+                    " \"" + jsonString + "\""
+        );
+        SwrveCommon.setSwrveCommon(this);
+
+        if (context instanceof Activity) {
+            this.context = new WeakReference<>(context.getApplicationContext());
+        } else {
+            this.context = new WeakReference<>(context);
+        }
+
+        SharedPreferences sp = this.context.get().getSharedPreferences("FILE", Context.MODE_PRIVATE);
+        if(null != jsonString) {
+            try {
+                initFromJSON(jsonString);
+                SharedPreferences.Editor editor = sp.edit();
+                editor.putString(LOG_TAG, jsonString);
+                editor.apply();
+            } catch (Exception e) {
+                jsonString = null;
+                SwrveLogger.e(LOG_TAG, "Error loading settings from JSON", e);
+            }
+        }
+
+        if(null == jsonString) {
+            try {
+                initFromJSON(sp.getString(LOG_TAG, ""));
+            } catch (Exception e) {
+                SwrveLogger.e(LOG_TAG, "Error loading Unity settings from shared prefs", e);
+            }
+        }
+
+        this.cacheDir = new File(getSwrveTemporaryPath());
+
+        sessionKey =
+            SwrveHelper.generateSessionToken(this.getApiKey(), this.getAppId(), getUserId()); // Generate session token
+    }
+
+    public void initFromJSON(String jsonString) {
+        SwrveLogger.d(
+            "UnitySwrveCommon:initFromJSON called with jsonString" +
+            " \"" + jsonString + "\""
+        );
+        Gson gson = new Gson();
+        this.currentDetails = gson.fromJson(jsonString, new TypeToken<Map<String, Object>>(){}.getType());
```

Comment:
Think a more descriptive name could be used instead of "currentDetails". Actually, do we even need this? If we're only storing 4 keys (SWRVE_TEMPORARY_PATH_KEY, SDK_VERSION_KEY, PREFAB_NAME_KEY, DEVICE_INFO_KEY) when save the values in 4 separate variables in this method. Makes for more easily readable and testable code.

Also, if you use <String, String> instead of <string, object>, won't this return a json string for DEVICE_INFO_KEY? Which we should be able to turn into a jsonobject by doing new JSONObject(json)......or something like that?